### ISSUE TITLE
Support baseUrl and paths resolution as implemented in TS 1.9

### ISSUE THREAD

**Issue by guncha** [s-panferov/awesome-typescript-loader#156]
I was trying to use `awesome-typescript-loader` with `typescript@next` to see if I could set it all up to avoid this whole `import {..} from "../../../../../dep` business and it seems like `baseUrl` option solves it, but only when not running through the loader.

When running with the loader, I'm getting an error:

```
Error: Cannot resolve module 'dep' in app
  at ResolutionError.Error (native)
  at new ResolutionError (node_modules/awesome-typescript-loader/dist.babel/deps.js:577:88)
  at node_modules/awesome-typescript-loader/dist.babel/deps.js:402:37
```

Does the loader have its own module resolution strategy? If so, can we extend it to support what's being discussed in https://github.com/Microsoft/TypeScript/issues/5039 or perhaps use the compiler's?

**Comment by s-panferov** [comment]
@guncha this error is about webpack and how it resolves modules. Webpack knows nothing about typescript, it uses basic `node-style` module resolution strategy. I need to think about how to connect webpack's resolution config with typescript's one, but right now there is a workaround:

* If you use `baseUrl`: add it to webpack's [resolve.root](https://webpack.github.io/docs/configuration.html#resolve-root)

* If you use `paths`: add it to webpack's [resolve.alias](https://webpack.github.io/docs/configuration.html#resolve-alias)

**Comment by guncha** [comment]
I see, that's very interesting. I had assumed that the loader tells webpack "_hey, add this files as dependencies_", but I guess it's the other way around.

**Comment by TheLarkInn** [comment]
@s-panferov how would you go about setting this in `alias` in webpack. (I've pasted below what I've been able to get to work in `ts-loader`. If it helps. 

`ts.config`
```
    "paths": {
      "@angular2-material/core/*": [
        "./core/*",
        "../core/*"
      ],
      "@angular2-material/*": [
        "./components/*",
        "../components/*"
      ]
    }

```
`webpack.config.js`

```
var components = [
    'button',
    'card',
    'checkbox',
    'grid-list',
    'icon',
    'input',
    'list',
    'progress-bar',
    'progress-circle',
    'radio',
    'sidenav',
    'slide-toggle',
    'button-toggle',
    'tabs',
    'toolbar'
];
/** Map relative paths to URLs. */
var aliasMap: any = {
    '@angular2-material/core': ngAppResolve('./src/core'),
};

components.forEach(function (name) {
  aliasMap[("@angular2-material/" + name)] = ngAppResolve("./src/components/" + name);
  return aliasMap[("@angular2-material/" + name)] = ngAppResolve("./src/components/" + name);
});

export const config = {
  resolve: {
    alias: aliasMap
  }
}
```

**Comment by s-panferov** [comment]
@TheLarkInn yes, aliases is the right way to go. I just need to think if it possible to set them automatically , using info from tsconfig.json.

I suppose we need something like

```
import { tsAliases } from `awesome-typescript-loader`

export const config = {
  resolve: {
     alias: tsAliases({ /* additional aliases */ })
  }
}
```

**Comment by TheLarkInn** [comment]
@s-panferov could you set it in compiler.plugin('run') or ('watch-run') or both.

**Comment by s-panferov** [comment]
@TheLarkInn I don't think so, because as far as I remember webpack creates resolver one time during init process. I need to dig into this.

**Comment by TheLarkInn** [comment]
@sokra do you have any suggestions for this? I'm not sure the _best_ way to handle aliasing like this.

**Comment by sokra** [comment]
You can write a resolver plugin for webpack, which takes care of resolving dependencies accoring to a `tsconfig` relative to the origin file.

**Comment by s-panferov** [comment]
@sokra ok, I'll try to implement this. Do you have any examples?

**Comment by TheLarkInn** [comment]
@s-panferov https://github.com/webpack/webpack/blob/master/lib/WebpackOptionsApply.js#L273-L286

Looks like how webpack reads aliases and applies them.

**Comment by sokra** [comment]
Here is the webpack alias plugin: https://github.com/webpack/enhanced-resolve/blob/master/lib/AliasPlugin.js

It's used here: https://github.com/webpack/enhanced-resolve/blob/master/lib/ResolverFactory.js#L155

---

So you can write your own alias plugin: `TsConfigAliasPlugin` and use it in the configuration:

``` js
resolve: {
  plugins: [
    new TsConfigAliasPlugin("described-resolve", "resolve")
  ]
}
```

**Comment by s-panferov** [comment]
Basic implementation landed in `v2.0.0-rc.15`.

```
var TsConfigPathsPlugin = require('awesome-typescript-loader').TsConfigPathsPlugin;

resolve: {
    plugins: [
        new TsConfigPathsPlugin()
    ]
}
```

Please try and say what you think.

**Comment by TheLarkInn** [comment]
Good timing I'll try right now. Great work.

**Comment by TheLarkInn** [comment]
`ReferenceError: ts is not defined
    at Object.readConfigFile (/Users/larkinfamily/Code/angular-cli/node_modules/awesome-typescript-loader/src/instance.ts:319:30)`

I didn't see `ts` defined anywhere in `instance.ts`

**Comment by TheLarkInn** [comment]
OH hold that though, I'm sory I set that up wrong. Let me try again.

**Comment by TheLarkInn** [comment]
@s-panferov I am still seeing that error. Just to confirm, I don't have to pass any arguments to the plugin, it likes in the code it already picked up and obtained from `instance`

**Comment by TheLarkInn** [comment]
Here is a [gist](https://gist.github.com/TheLarkInn/954f866378d16229846beecb4da1a070) of the configuration if this helps.

**Comment by s-panferov** [comment]
@TheLarkInn just to confirm, you don't have a `tsconfig.json` file in your project?

**Comment by s-panferov** [comment]
I've fixed the error in `2.0.0-rc.16`, but it appeared only when you don't have tsconfig.json.

This is right config for your app:

```
new TsConfigPathsPlugin({
   tsconfig: ngAppResolve('./src/demo-app/tsconfig.json'),
})
```

Unfortunately, the plugin can't communicate with the loader, so you have to specify `tsconfig` option twice for now.

**Comment by TheLarkInn** [comment]
This is for a situation where angular CLI I which generates projects. so the tsconfig is not at the root execution path it's the projects patg

**Comment by TheLarkInn** [comment]
Thank you that's no problem at all. I'll let you know here in a bit, sorry for this edge-casey kind of stuff.

**Comment by s-panferov** [comment]
@TheLarkInn could you please confirm that everything works?

**Comment by TheLarkInn** [comment]
@s-panferov Yup. I'll take a look at it again today. And let you know what I find.

**Comment by guncha** [comment]
Nice, I can confirm that it works with `baseUrl` and `paths` options in `tsconfig.json`. A few things that weren't immediately obvious to me:

1. This is a webpack 2.0 plugin and the configuration has no effect when used with 1.x.
2. I had to add `*.ts` and `*.tsx` to `resolve.extensions`. That seems a little odd, but I think I understand the underlying reason for it. Perhaps the plugin could ensure that those are added?

Either way, awesome work. Can't wait to use this. Now if the editors could catch up to 1.9.x, that'd be great.

**Comment by TheLarkInn** [comment]
My edge case project is really hard to test this feature and tell you reliably if it's working so I'm going to create a test project and test this feature more in depth.

**Comment by TheLarkInn** [comment]
@s-panferov I had to back port an isolated hack of this to v1.1.1 but when I did I could confirm it was working (did have some a-t-loader diagnostic errors because a-t-loader 's Resolver was failing to find 1 or two files). In regards to the back port there seems to be some strange path resolution issues outside of this use. 

Second when `baseUrl:''` it needs to default to './' or path.dirname(configFilePath). Sorry for the delay with this I've been cross coding like 5 libraries webpack and angular-cli being two of them.
